adr/ADR-15.md

```diff
@@ -0,0 +1,235 @@
+- An optional stream name
+- An optional consumer name
+- An optional JS consumer configuration (see below)
+- A Queue name if this subscription is meant to be a queue subscription
+- Indication if this is a "Push" or "Pull" subscription
+- Handlers for receiving messages asynchronously, whatever that means in the library language
+
+Some misconfiguration should be checked by the subscribe API and return an error outright.
+
+For instance:
+- If a queue name is provided, configuring heartbeat or flow control is a mistake, since the server would send those message to random members.
+- If no subject is provided, along with no stream name, the library won't be able to locate/create the consumer.
+- For pull subscriptions ack policy of "none" or "all" is an error.
+
+At the time of this writing, the consumer configuration object looks like this (ordred alphabetically, not as seen in js.go):
+
+```go
+type ConsumerConfig struct {
+	AckPolicy       AckPolicy     `json:"ack_policy"`
+	AckWait         time.Duration `json:"ack_wait,omitempty"`
+	DeliverGroup    string        `json:"deliver_group,omitempty"`
+	DeliverPolicy   DeliverPolicy `json:"deliver_policy"`
+	DeliverSubject  string        `json:"deliver_subject,omitempty"`
+	Description     string        `json:"description,omitempty"`
+	Durable         string        `json:"durable_name,omitempty"`
+	FilterSubject   string        `json:"filter_subject,omitempty"`
+	FlowControl     bool          `json:"flow_control,omitempty"`
+	Heartbeat       time.Duration `json:"idle_heartbeat,omitempty"`
+	MaxAckPending   int           `json:"max_ack_pending,omitempty"`
+	MaxDeliver      int           `json:"max_deliver,omitempty"`
+	MaxWaiting      int           `json:"max_waiting,omitempty"`	// server returns error if provided if a DeliverSubject is provided
+	OptStartSeq     uint64        `json:"opt_start_seq,omitempty"`	// server returns error if provided when Deliver Policy is not by_start_sequence
+	OptStartTime    *time.Time    `json:"opt_start_time,omitempty"`	// server returns error if provided when Deliver Policy is not by_start_time
+	RateLimit       uint64        `json:"rate_limit_bps,omitempty"` // Bits per sec
+	ReplayPolicy    ReplayPolicy  `json:"replay_policy"`
+	SampleFrequency string        `json:"sample_freq,omitempty"`
+}
+```
+
+#### Stream name
+
+When no stream is specified, the `subject` to the subscribe API calls is required. An error will be returned if it is not the case.
+The library will use the subject provided as a way to find out which stream this subscription is for. A request is sent to the server on the `<JS prefix>.STREAM.NAMES` subject with a JSON content `{"subject":"<subject here>"}`. If the response (a list of stream names) is positive and contains a single entry, then the library will use this stream name, otherwise an error indicating that there is no matching stream name should be returned.
+
+If the stream name is specified, then the `subject` is possibly not needed.
```

Comment:
should clients assume a subject equal to the stream name?

Reply:
Now that we have removed the SubjectIsDelivery, things change a bit, so I may update this. If it is a pull subscription, if the stream is specified, this subject will really not be needed if the consumer already exists. But for non pull, or when creating the JS consumer, it is needed because we set FilterSubject with the given subject, or, for the lookup we verify that if a filter subject is set, it matches the user given subject.
So I will see how I update that.